### ISSUE TITLE
Support for Pure Cool + Humidifier

### ISSUE THREAD

**Issue by charliebird** [lukasroegner/homebridge-dyson-pure-cool#58]
Dyson has launched a new Pure Cool + Humidifier which combines the existing Pure Cool tower with a humidifier. I guess this plugin can control the product like a normal Pure Cool, but I was wondering if it’s possible to add support for the Humidifier

**Comment by lukasroegner** [comment]
As soon as it uses the same protocol, it can be supported. I‘m not planning to buy one, so I need help from someone you owns such a device.

**Comment by charliebird** [comment]
Thanks for the quick reply. I ordered one arriving on Tuesday so I can help

**Comment by olliox** [comment]
I tried the dyson pure cool humidifier...that‘s what I got:
[DysonPureCoolPlatform] Device with serial number C4H-EU-N......A  not added, as it is not supported by this plugin. Product type: 358
The serial was of course the right one in the original log...just changed it here

**Comment by lukasroegner** [comment]
Support for the new humidifier has to be explicitly implemented, I'll create a new branch for the development and notify you.

**Comment by lukasroegner** [comment]
For those of you who want to help me with supporting the PH01:

1. Please install the beta version via:
```
npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01
```

2. Configure the plugin as stated in the README.

3. Start homebridge in debug mode (parameter `-D`)

4. Provide me with log entries for MQTT messages of type `CURRENT-STATE` and `ENVIRONMENTAL-STATE-CHANGE`

I'll go through the MQTT messages and try to find out, which properties have to be mapped in the first round.

**Comment by charliebird** [comment]
Unfortunately, Dyson messed up my order, and sent me the wrong product and now I'm not gonna receive the PH01 until mid June, so I won't be able to help before that...

**Comment by lukasroegner** [comment]
Thank you @boldiebold! I extracted the `product-state` values from your logs and compared them to the state that is available on the TP04 and DP04 devices. Now, we have to identify properties and values that are important for the humidifier. 

Please update the plugin via NPM:
```
npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01
```

For the next step, it is no longer needed to start homebridge in debug mode.
You will see "STATE" entries in the log, where property names (4 letters) and their values are printed out. This should help you to identify the meaning of some of the properties. Please let homebridge run and change a setting in the Dyson app. You should immediately see new log entries in homebridge. Now, check for changes of the state by comparing it to the previous log entries.

Here is a table of properties that exist. It contains all of my findings:

| STATE | PH01 | TP04 | POSSIBLE VALUES | DESCRIPTION |
| --- | --- | --- | --- | --- |
| fpwr | X | X | ON, OFF | Fan power |
| auto | X | X | ON, OFF | Auto mode |
| oscs | X | X | ON, OFF | ? |
| oson | X | X | ON, OFF | Oscillation mode |
| nmod | X | X | ON, OFF | Night mode |
| rhtm | X | X | ON, OFF | ? |
| fnst | X | X | ON, OFF | Fan state |
| ercd | X | X | ? | ? |
| wacd | X | X | NONE, ? | ? |
| nmdv | X | X | 0040, ? | Night mode fan speed? |
| fnsp | X | X | 0001 - 0010, AUTO | Fan speed |
| bril | X | X | 0002, ? | ? |
| corf | X | X | ON, ? | ? |
| cflr | X | X | 0000 - 0100, INV | Carbon filter life in percent |
| hflr | X | X | 0000 - 0100, INV | HEPA filter life in percent |
| cflt | X | X | NONE, CARF, ? | Carbon filter type? |
| hflt | X | X | NONE, GCOM, GHEP | HEPA filter type? |
| sltm | X | X | OFF, ? | ? |
| osal | X | X | 01XX, ? | Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app. |
| osau | X | X | 01XX, ? | Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app. |
| ancp | X | X | CUST, BRZE, ? | Determines the type of oscillation. CUST is used for custom angle for TP04 and DPß4. BRZE could be the "breeze" mode of PH01? |
| hume | X |  | HUMD, ? | ? |
| haut | X |  | ON, ? | Could be AUTO mode for humidity? |
| humt | X |  | 0040, ? | Could be target humidity? |
| rect | X |  | 0040, ? | ? |
| msta | X |  | OFF, ? | ? |
| clcr | X |  | CLNO, ? | ? |
| cdrr | X |  | 0060, ? | ? |
| cltr | X |  | 1218, ? | ? |
| wath | X |  | 1350, ? | Could be remaining water? |
| psta | X |  | OFF, ? | ? |
| fdir | X | X | OFF, ON | Jet Focus |

Our goal should be to identify at least the properties that are only available on the PH01. Please post your findings here, I'll update the table.

**Comment by boldiebold** [comment]
Hi Lukas,
Here are the additions I recognized so far.





STATE   PH01    TP04    POSSIBLE VALUES DESCRIPTION
fpwr    X       X       ON, OFF Fan power
auto    X       X       ON, OFF Auto mode
oscs    X       X       ON, OFF, idle    Idle when water tank is not inside or empty
oson    X       X       ON, OFF Oscillation mode
nmod    X       X       ON, OFF Night mode
rhtm    X       X       ON, OFF ?
fnst    X       X       ON, OFF FAN     Fan state
ercd    X       X       31U2 ? 30u1 14II        Error Code
wacd    X       X       NONE, ? ?
nmdv    X       X       0040, ? 0004    Night mode fan speed?
fnsp    X       X       0001 - 0010, AUTO       Fan speed
bril    X       X       0002, ? ?
corf    X       X       ON, ?   ?
cflr    X       X       0000 - 0100, INV        Carbon filter life in percent
hflr    X       X       0000 - 0100, INV        HEPA filter life in percent
cflt    X       X       NONE, CARF, ?   Carbon filter type?
hflt    X       X       NONE, GCOM, GHEP        HEPA filter type?
sltm    X       X       OFF, ?  ?
osal    X       X       01XX, ? 0159 0178  0161 0169 0180 0205  Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app. (No customising)
osau    X       X       01XX, ? 0159 0178  0161 0169 0180 0205  Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app. (No customising)
ancp    X       X       CUST, BRZE, ?0090 0045          Determines the type of oscillation. CUST is used for custom angle for TP04 and DPß4. BRZE could be the "breeze" mode of PH01?
hume    X               HUMD, ?off      Humidity

haut    X               ON, off AUTO mode for humidity? Yes
humt    X               0040 - 0070     Could be target humidity Yes
rect    X               0040, 0035      ?
msta    X               OFF, ? HUMD     Status Humidity on / off
clcr    X               CLNO, ? ?
cdrr    X               0060, ? ?
cltr    X               1218, ?1202 1201        ?
wath    X               1350, ? Could be remaining water? 1350 auch wenn Wassertank leer
psta    X               OFF, init, clng Status Air Cleaning
fdir    X       X       OFF, ON Jet Focus


Anfang der weitergeleiteten Nachricht:

Von: Lukas Rögner <notifications@github.com<mailto:notifications@github.com>>
Betreff: Aw: [lukasroegner/homebridge-dyson-pure-cool] Support for Pure Cool + Humidifier (#58)
Datum: 24. April 2020 um 08:51:58 MESZ
An: lukasroegner/homebridge-dyson-pure-cool <homebridge-dyson-pure-cool@noreply.github.com<mailto:homebridge-dyson-pure-cool@noreply.github.com>>
Kopie: boldiebold <owuerker@outlook.com<mailto:owuerker@outlook.com>>, Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Antwort an: lukasroegner/homebridge-dyson-pure-cool <reply@reply.github.com<mailto:reply@reply.github.com>>



Thank you @boldiebold<https://github.com/boldiebold>! I extracted the product-state values from your logs and compared them to the state that is available on the TP04 and DP04 devices. Now, we have to identify properties and values that are important for the humidifier.

Please update the plugin via NPM:

npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01


For the next step, it is no longer needed to start homebridge in debug mode.
You will see "STATE" entries in the log, where property names (4 letters) and their values are printed out. This should help you to identify the meaning of some of the properties. Please let homebridge run and change a setting in the Dyson app. You should immediately see new log entries in homebridge. Now, check for changes of the state by comparing it to the previous log entries.

Here is a table of properties that exist. It contains all of my findings:

STATE   PH01    TP04    POSSIBLE VALUES DESCRIPTION
fpwr    X       X       ON, OFF Fan power
auto    X       X       ON, OFF Auto mode
oscs    X       X       ON, OFF ?
oson    X       X       ON, OFF Oscillation mode
nmod    X       X       ON, OFF Night mode
rhtm    X       X       ON, OFF ?
fnst    X       X       ON, OFF Fan state
ercd    X       X       ?       ?
wacd    X       X       NONE, ? ?
nmdv    X       X       0040, ? Night mode fan speed?
fnsp    X       X       0001 - 0010, AUTO       Fan speed
bril    X       X       0002, ? ?
corf    X       X       ON, ?   ?
cflr    X       X       0000 - 0100, INV        Carbon filter life in percent
hflr    X       X       0000 - 0100, INV        HEPA filter life in percent
cflt    X       X       NONE, CARF, ?   Carbon filter type?
hflt    X       X       NONE, GCOM, GHEP        HEPA filter type?
sltm    X       X       OFF, ?  ?
osal    X       X       01XX, ? Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app.
osau    X       X       01XX, ? Determines the oscillation angle? Can be customized for TP04 and DP04 in the Dyson app.
ancp    X       X       CUST, BRZE, ?   Determines the type of oscillation. CUST is used for custom angle for TP04 and DPß4. BRZE could be the "breeze" mode of PH01?
hume    X               HUMD, ? ?
haut    X               ON, ?   Could be AUTO mode for humidity?
humt    X               0040, ? Could be target humidity?
rect    X               0040, ? ?
msta    X               OFF, ?  ?
clcr    X               CLNO, ? ?
cdrr    X               0060, ? ?
cltr    X               1218, ? ?
wath    X               1350, ? Could be remaining water?
psta    X               OFF, ?  ?
fdir    X       X       OFF, ON Jet Focus

Our goal should be to identify at least the properties that are only available on the PH01. Please post your findings here, I'll update the table.

**Comment by lukasroegner** [comment]
I updated the table in my post above. There should be enough information for me to try a first implementation of the humidifier feature of the device.

**Comment by Christian-CFP** [comment]
If help is needed let me know.
Own a PH1 and can also test.
Greetz Christian

**Comment by Christian-CFP** [comment]
I user hoobs and i cant install it with
`npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01`

**Comment by lukasroegner** [comment]
I'm not using Hoobs and thus cannot tell you. I would also not recommend to use Hoobs for anything that is not stable. You have to wait until the implementation is completed and other people have tested it.

**Comment by Christian-CFP** [comment]
Hey Lukas,
when do you expect to release it?

**Comment by lukasroegner** [comment]
It will be merged into master when the implementation is finished, tested and mostly bug-free.

@boldiebold New version to test is on the `ph01` branch. Please update via

```
npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01
```

New: new accessory (humidifier) with support for:
* Current State: OFF/IDLE/HUMIDIFYING
* Target State: OFF/AUTO/MANUAL
* Target humidity (40% - 70%)

Please try it out and report issues with the implementation.

**Comment by boldiebold** [comment]
Hi Lukas,

Great stuff. Works fine. Don’t find any issues yet.
Just some German expression I read at some points  ENTfeuchten instead of BEfeuchten.
Perhaps you could change it, but that’s minor cosmetic ;-)

Thanks for your great work.


Anfang der weitergeleiteten Nachricht:

Von: Lukas Rögner <notifications@github.com<mailto:notifications@github.com>>
Betreff: Aw: [lukasroegner/homebridge-dyson-pure-cool] Support for Pure Cool + Humidifier (#58)
Datum: 1. Mai 2020 um 15:54:50 MESZ
An: lukasroegner/homebridge-dyson-pure-cool <homebridge-dyson-pure-cool@noreply.github.com<mailto:homebridge-dyson-pure-cool@noreply.github.com>>
Kopie: boldiebold <owuerker@outlook.com<mailto:owuerker@outlook.com>>, Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Antwort an:



It will be merged into master when the implementation is finished, tested and mostly bug-free.

@boldiebold<https://github.com/boldiebold> New version to test is on the ph01 branch. Please update via

npm install -g https://github.com/lukasroegner/homebridge-dyson-pure-cool.git#ph01


New: new accessory (humidifier) with support for:

  *   Current State: OFF/IDLE/HUMIDIFYING
  *   Target State: OFF/AUTO/MANUAL
  *   Target humidity (40% - 70%)

Please try it out and report issues with the implementation.

**Comment by lukasroegner** [comment]
@boldiebold Thank you for testing it out!

Plugins don't provide any UI or wording, this is all handled by the Home app...did you read ENTfeuchten somewhere? That might be a bug in the plugin. HomeKit provides a "HumidifierDehumidifier" service, which I used for the implementation. I disabled the dehumidifier part (as the PH01 is just a humidifier), maybe I forgot to disable it somewhere?

**Comment by boldiebold** [comment]
[cid:94EAC54A-06D4-4C65-9D1E-B845B219DFE5-L0-001]

Am 02.05.2020 um 17:02 schrieb Lukas Rögner <notifications@github.com>:

﻿

@boldiebold<https://github.com/boldiebold> Thank you for testing it out!

Plugins don't provide any UI or wording, this is all handled by the Home app...did you read ENTfeuchten somewhere? That might be a bug in the plugin. HomeKit provides a "HumidifierDehumidifier" service, which I used for the implementation. I disabled the dehumidifier part (as the PH01 is just a humidifier), maybe I forgot to disable it somewhere?

**Comment by lukasroegner** [comment]
@boldiebold Can you upload the screenshot directly to GitHub? It's not loaded properly from your email.

**Comment by boldiebold** [comment]
![IMG_5954](https://user-images.githubusercontent.com/52777334/80868156-f7e32e80-8c98-11ea-97e6-6f156727e78b.jpeg)

It's out of Home+ 4

**Comment by lukasroegner** [comment]
I see, Home+ 4 calls the device a "Luftentfeuchter", however, the modes are all correct ("Befeuchten", "Schwellenwert für Befeuchten"). This is fine, I think that Home+ does not have special names for the HumidifierDehumidifier service depending on the supported features.

**Comment by lukasroegner** [comment]
Support for PH01 is not in master branch and new version published on NPM (see #74).

**Comment by boldiebold** [comment]
Seems That Activating the humidifier will not work. Perhaps anybody else can test it.
![IMG_5954](https://user-images.githubusercontent.com/52777334/80917817-9c30a800-8d61-11ea-833a-2dd2cf1b83b3.jpeg)
